### PR TITLE
dns: add dns.rcode keyword v5

### PR DESCRIPTION
**Pull request by hadiqaalamdar**
Feature #6621

- [x] I have read the contributing guide lines at https://docs.suricata.io/en/latest/devguide/codebase/contributing/contribution-process.html
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata.io/about/contribution-agreement/
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/issues/6621) ticket: https://redmine.openinfosecfoundation.org/issues/6621

Previous PR: https://github.com/OISF/suricata/pull/10150

Describe changes:
- modified the code to implement `DetectUintData<u8>` instead of simple u8
- SV tests are running successfully
- cherry-picked changes from https://github.com/OISF/suricata/pull/10142/commits/1fc18c11b95864a4cd7bd1c81866055460c4377d to the etc/schema.json file to fix the schema errors
- cherry-picked changes from this PR https://github.com/OISF/suricata/pull/10073 which modified opcode to  implement `DetectUintData<u8>`. This PR had code that I would have had to replicate, especially `uint.rs` file changes.
- Added documentation for rcode in `doc/userguide/rules/dns-keywords.rst` file.
- successfully ran the following tests:
- from rust dir: make check
- from rust dir: cargo test
- from rust dir: cargo clippy --all-features
- from rust dir: rustfmt src/dns/*
- from suricata dir: ./scripts/clang-format.sh check-branch
- from suricata dir: ./scripts/clang-format.sh branch
- from suricata dir running ./src/suricata -u showed that one unit test had failed. 
- from suricata dir running ./src/suricata -u -U dns showed 7 tests passed.

SV_BRANCH=https://github.com/OISF/suricata-verify/pull/1578

### CODE REVIEW

File: doc/userguide/rules/dns-keywords.rst [hunk 1]
```diff
@@ -46,6 +48,47 @@ Match on DNS requests where the **opcode** is NOT 0::
...
+Match on DNS requests where the **opcode** is between 7 and 15, exclusively:
+
+  dns.opcode:7-15;
+
+Match on DNS requests where the **opcode** is not between 7 and 15:
+
+  dns.opcode:!7-15;
+
+dns.rcode
+----------
```
Comment (jlucovsky) [inline doc/userguide/rules/dns-keywords.rst:60]: nit: make sure that the number of `-` characters matches that in the text above. In this case, there's one extra `-`

File: doc/userguide/rules/dns-keywords.rst [hunk 2]
```diff
@@ -46,6 +48,47 @@ Match on DNS requests where the **opcode** is NOT 0::
...
+
+Match on DNS requests and responses with **rcode** 4::
+
+  dns.rcode:4;
+
+Match on DNS requests where the **rcode** is NOT 0::
+
+  dns.rcode:!0;
+
+Match on DNS requests where the **rcode** is between 7 and 15, exclusively:
```
Comment (jlucovsky) [inline doc/userguide/rules/dns-keywords.rst:84]: Are 7 and 15 included? If so, suggest "between 7 and 15, inclusively"; exclusive implies something differrent.

[7-15] -- inclusive
[7-15) -- exclusive (15 not included)

File: rust/src/dns/detect.rs [hunk 3]
```diff
@@ -76,116 +42,219 @@ pub extern "C" fn rs_dns_opcode_match(
...
-pub unsafe extern "C" fn rs_detect_dns_opcode_parse(carg: *const c_char) -> *mut c_void {
-    if carg.is_null() {
-        return std::ptr::null_mut();
-    }
-    let arg = match CStr::from_ptr(carg).to_str() {
-        Ok(arg) => arg,
-        _ => {
-            return std::ptr::null_mut();
+pub extern "C" fn rs_dns_rcode_match(
+    tx: &mut DNSTransaction, detect: &mut DetectUintData<u8>, flags: u8,
```
Comment (jufajardini) [inline rust/src/dns/detect.rs:55]: Could you please add documentation for this function?

File: rust/src/dns/detect.rs [hunk 4]
```diff
@@ -76,116 +42,219 @@ pub extern "C" fn rs_dns_opcode_match(
...
+        // Not to server or to client??
+        return 0;
     };
 
-    match parse_opcode(arg) {
-        Ok(detect) => Box::into_raw(Box::new(detect)) as *mut _,
-        Err(_) => std::ptr::null_mut(),
-    }
-}
+    let rcode = (header_flags & 0xf) as u8;
```
Comment (jufajardini) [inline rust/src/dns/detect.rs:74]: Honest question: why do we need the bitmask for rcode?

**Review by jufajardini — CHANGES_REQUESTED** [review]
Getting closer and closer to the finish line! :)

It's a bit difficult to check what changes were introduced by you due to the commit history at the moment, but tried to add some useful feedback inline. :)

**Comment by jufajardini** [comment]
Jason's commit got merged, can be a good idea to have a new version rebased, to clean up the commit history a bit more :)

**Comment by catenacyber** [comment]
I think you should remove the commit from https://github.com/OISF/suricata/pull/10073 from this PR. 

These are 2 independent changes, right ? (even if they are close to each other)

**Comment by catenacyber** [comment]
> modified the code to implement DetectUintData<u8> instead of simple u8

Looks cool :-)

**Comment by hadiqaalamdar** [comment]
> I think you should remove the commit from #10073 from this PR.
> 
> These are 2 independent changes, right ? (even if they are close to each other)
I had to copy some of the changes, e.g changes in uint.rs, file you made in #10073. Is it okay if I cherrypick the 2 commits from #10073 or do you think that I should copy them in my PR without cherrypicking?

**Comment by catenacyber** [comment]
> I had to copy some of the changes, e.g changes in `uint.rs` file, you made in #10073.

I do not think you have to do that...
This PR should not change `uint.rs`

Why do you think you need this ?

>  Is it okay if I cherrypick the 2 commits from #10073 or do you think that I should copy them in my PR without cherrypicking and without changing the code for `opcode`?

I updated #10073 (there is a new PR)
But I think you can use `master` branch instead of this PR as a base for your work...

**Comment by hadiqaalamdar** [comment]
New PR: https://github.com/OISF/suricata/pull/10233

**Pull request closed without merging**